src/openloco/windows/CompanyWindow.cpp

```diff
@@ -0,0 +1,1985 @@
+            { challenge::widgets, widx::tab_challenge, &challenge::events, &challenge::enabledWidgets, &challenge::windowSize }
+        };
+
+        static void initEvents()
+        {
+            status::initEvents();
+            details::initEvents();
+            colour_scheme::initEvents();
+            finances::initEvents();
+            cargo_delivered::initEvents();
+            challenge::initEvents();
+        }
+
+        static void switchTabWidgets(window* self)
+        {
+            self->activated_widgets = 0;
+
+            static widget_t* widgetCollectionsByTabId[] = {
+                status::widgets,
+                details::widgets,
+                colour_scheme::widgets,
+                finances::widgets,
+                cargo_delivered::widgets,
+                challenge::widgets,
+            };
+
+            widget_t* newWidgets = widgetCollectionsByTabId[self->current_tab];
+            if (self->widgets != newWidgets)
+            {
+                self->widgets = newWidgets;
+                // self->init_scroll_widgets();
+            }
+
+            static const widx tabWidgetIdxByTabId[] = {
+                tab_status,
+                tab_details,
+                tab_colour_scheme,
+                tab_finances,
+                tab_cargo_delivered,
+                tab_challenge,
+            };
+
+            self->activated_widgets &= ~((1 << tab_status) | (1 << tab_details) | (1 << tab_colour_scheme) | (1 << tab_finances) | (1 << tab_cargo_delivered) | (1 << tab_challenge));
+            self->activated_widgets |= (1ULL << tabWidgetIdxByTabId[self->current_tab]);
+        }
+
+        // 0x0043230B
+        static void switchTab(window* self, widget_index widgetIndex)
+        {
+            if (input::is_tool_active(self->type, self->number))
+                input::cancel_tool();
+
+            textinput::sub_4CE6C9(self->type, self->number);
+
+            self->current_tab = widgetIndex - widx::tab_status;
+            self->frame_no = 0;
+            self->flags &= ~(window_flags::flag_16);
+
+            if (self->viewports[0] != nullptr)
+            {
+                self->viewports[0]->width = 0;
+                self->viewports[0] = nullptr;
+            }
+
+            auto tabIndex = widgetIndex - widx::tab_status;
+            auto tabInfo = tabInformationByTabOffset[tabIndex];
+
+            self->enabled_widgets = *tabInfo.enabledWidgets;
+            self->holdable_widgets = 0;
+            self->event_handlers = tabInfo.events;
+            self->activated_widgets = 0;
+            self->widgets = tabInfo.widgets;
+
+            if (tabInfo.widgetIndex == widx::tab_finances)
+                self->holdable_widgets = finances::holdableWidgets;
+
+            common::disableChallengeTab(self);
+
+            self->set_size(*tabInfo.windowSize);
```

Comment:
you are missing a `self->invalidate()` prior to setting the size. It causes drawing issues due to the lack of marking of dirty block.